Update config model docs (#15608)
DataDog/integrations-core: docs/developer/meta/config-models.md

There are 2 possible models:

- `InstanceConfig` (ID: `instance`) that corresponds to a check's entry in the `instances` section
- `SharedConfig` (ID: `shared`) that corresponds to the `init_config` section that is shared by all instances

All models are defined in `<ID>.py` and are available for import directly under `config_models`.

## Default values

The default values for optional settings are populated in `defaults.py` and are derived from the
[value](config-specs.md#values) property of config spec options. The precedence is the `default` key
followed by the `example` key (if it appears to represent a real value rather than an illustrative example

> and the `type` is a primitive). In all other cases the default will be `None` and there will be no default

Reviewer: ```suggestion
and the `type` is a primitive). In all other cases, the default is `None`. There is no default
```

Author: This isn't quite right because the intent is to express that there is no function when there is no default

Reviewer: Oh, I see, i misunderstood. Let me edit again. Thanks for explaining!

getter function.

## Validation

> The validation of fields for every model occurs in 3 high-level stages.

Reviewer: ```suggestion
The validation of fields for every model occurs in three high-level stages, as described in this section.
```

If such a validator exists in `validators.py`, then it is called once with the raw config that was supplied by the user.
The returned mapping is used as the input config for the subsequent stages.

### Field

The value of each field goes through the following steps.

#### Default value population

If a field was not supplied by the user nor during the [initialization stage](#initial), then its default value is
taken from `defaults.py`. This stage is skipped for required fields.

#### Custom field validators

The contents of `validators.py` are entirely custom and contain functions to perform extra validation if necessary.

```python
def <ID>_<OPTION_NAME>(value: Any, *, field: pydantic.fields.FieldInfo, **kwargs) -> Any:
    ...
```

Such validators are called for the appropriate field of the proper model. The returned value is used as the
new value of the option for the subsequent stages.

> !!! note

Reviewer: ```suggestion
**Note**: This only occurs if the option was supplied by the user.
```

Author: This is a feature of our documentation framework

Reviewer: I see, deleting my suggestions here.

>     This only occurs if the option was supplied by the user.

Reviewer: ```suggestion
```

#### Pre-defined field validators

A `validators` key under the [value](https://datadoghq.dev/integrations-core/meta/config-specs/#values) property of config

> spec options is considered. Every entry will refer to a relative import path to a [field validator](#custom-field-validators)

Reviewer: ```suggestion
spec options is considered. Every entry refers to a relative import path to a [field validator](#custom-field-validators)
```

under `datadog_checks.base.utils.models.validation` and is executed in the defined order.

> !!! note

Reviewer: ```suggestion
**Note**: This only occurs if the option was supplied by the user.
```

Author: This is a feature of our documentation framework

>     This only occurs if the option was supplied by the user.

Reviewer: ```suggestion
```

#### Conversion to immutable types

Every `list` is converted to `tuple` and every `dict` is converted to [types.MappingProxyType](https://docs.python.org/3/library/types.html#types.MappingProxyType).

> !!! note

Reviewer: ```suggestion
```

Author: This is a feature of our documentation framework

>     A field or nested field would only be a `dict` when it is defined as a mapping with arbitrary keys. Otherwise, it would be a model with its own properties as usual.

Reviewer: ```suggestion
**Note**: A field or nested field would only be a `dict` when it is defined as a mapping with arbitrary keys. Otherwise, it would be a model with its own properties as usual.
```

### Final

```python
def check_<ID>(model: pydantic.BaseModel) -> pydantic.BaseModel:
    ...
```

> If such a validator exists in `validators.py`, then it is called with the final constructed model. At this point it cannot

Reviewer: ```suggestion
If such a validator exists in `validators.py`, then it is called with the final constructed model. At this point, it cannot
```

> be mutated so you may only raise errors.

Reviewer: ```suggestion
be mutated, so you can only raise errors.
```
